driver-core/src/test/java/com/datastax/driver/core/CCMTestsSupport.java

```diff
@@ -0,0 +1,898 @@
+            this.numberOfNodes = numberOfNodes;
+        }
+
+        private CCMTestContext createAndStartTestContext() {
+            CCMTestContext ccmTestContext = new CCMTestContext(ccmBridgeBuilder, numberOfNodes);
+            ccmTestContext.start();
+            return ccmTestContext;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            CCMTestContextKey that = (CCMTestContextKey) o;
+            return ccmBridgeBuilder.equals(that.ccmBridgeBuilder) && Arrays.equals(numberOfNodes, that.numberOfNodes);
+
+        }
+
+        @Override
+        public int hashCode() {
+            int result = ccmBridgeBuilder.hashCode();
+            result = 31 * result + Arrays.hashCode(numberOfNodes);
+            return result;
+        }
+    }
+
+    private static class CCMTestContext {
+
+        private static final AtomicInteger COUNTER = new AtomicInteger(1);
+
+        private static final AtomicInteger NEXT_PORT = new AtomicInteger(5000);
+
+        private static final ReentrantLock PORT_LOCK = new ReentrantLock();
+
+        private final int id;
+
+        private final CCMBridge.Builder ccmBridgeBuilder;
+
+        private final int[] numberOfNodes;
+
+        private CCMBridge ccmBridge;
+
+        private List<CCMNode> nodes;
+
+        private boolean keepLogs = false;
+
+        private boolean stopped = false;
+
+        public CCMTestContext(CCMBridge.Builder ccmBridgeBuilder, int... numberOfNodes) {
+            this.id = COUNTER.getAndIncrement();
+            this.ccmBridgeBuilder = ccmBridgeBuilder;
+            this.numberOfNodes = numberOfNodes;
+            this.keepLogs = false;
+        }
+
+        private void start() {
+            LOGGER.debug("Starting " + this);
+            // force no start and no nodes
+            ccmBridge = ccmBridgeBuilder
+                    .withoutNodes()
+                    .notStarted()
+                    .build();
+            Runtime.getRuntime().addShutdownHook(new Thread() {
+                @Override
+                public void run() {
+                    CCMTestContext.this.stop();
+                }
+            });
+            Properties props = new Properties();
+            nodes = new ArrayList<CCMNode>();
+            boolean multiDc = numberOfNodes.length > 1;
+            int total = 1;
+            for (int nodesPerDc : numberOfNodes) {
+                total *= nodesPerDc;
+            }
+            PORT_LOCK.lock();
+            try {
+                // storage port must be the same for all nodes in CCM cluster;
+                // find first available for first ip, and hope it will work for all ips
+                int storagePort = nextPort(1);
+                List<TokenRange> ranges = new TokenRange(FACTORY.minToken(), FACTORY.minToken(), FACTORY).splitEvenly(total);
+                int n = 1;
```

Comment:
Even though nodes are in different data centers, they are still on the same ring.  Because of that, no node can have the same token.  They are offset by 100 (instead of 1 for example) to allow for nodes being replaced (you usually will bootstrap a new node at the node you are replacing token + 1) and offsetting them consistently by 100 from the other datacenter keeps them evenly distributed.